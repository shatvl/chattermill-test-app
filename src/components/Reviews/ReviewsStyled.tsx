import styled from 'styled-components';
import { Button } from '../Button';

export const ReviewList = styled.ul`
  padding: 0;
  margin: 0;
  list-style: none;
  margin-top: 1.25rem;
`;

export const ReviewListItem = styled.li`
  display: flex;
  align-items: center;
  justify-content: center;
  margin: 0 0 1rem;
`;

export const NotFoundContent = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  margin-top: 2rem;
  font-size: 1.125rem;
  ${Button} {
    margin-left: 0.5rem;
  }
`;
