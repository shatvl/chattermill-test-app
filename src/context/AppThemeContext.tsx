import React, { FC, createContext, useContext } from 'react';
import { DefaultTheme, ThemeProvider } from 'styled-components';

interface AppThemeContextProps {}

export interface AppTheme extends DefaultTheme {
  colors: {
    [key: string]: string;
  };
  zIndex: {
    [key: string]: number;
  };
  fonts: {
    header: string;
    main: string;
  };
  borderWidth: string;
}

// global theme for app, TODO: dark / light mode
export const getTheme = (): AppTheme => ({
  zIndex: {
    xs: 1,
    s: 2,
    m: 3,
    lg: 4,
    xlg: 5,
    xxlg: 6,
  },
  colors: {
    background: '#fff',
    primary: '#ff8489',
    primaryLight: 'rgba(255, 132, 136, .2)',
    primaryMedium: 'rgba(255, 132, 136, .5)',
    red: '#d65942',
    gray: '#c4c9cc',
    grayLight: 'e3e3e3',
    text: '#2e3939',
    textLighter: '#2e3939',
    textSecondary: '#707579',
    textMeta: '#bfbfbf',
    black: '#000000',
    white: '#ffffff',
  },
  fonts: {
    header: 'Hind Siliguri',
    main: 'Hind Siliguri',
  },
  borderWidth: '0.0625rem',
});

export const AppThemeContext = createContext<AppThemeContextProps>({});

export const AppThemeProvider: FC = ({ children }) => (
  //we could implement dark / light theme here, or user specific styles base on user session
  <AppThemeContext.Provider value={{}}>
    <ThemeProvider theme={getTheme()}>{children}</ThemeProvider>
  </AppThemeContext.Provider>
);

export const usePortalAppThemeContext = () => useContext(AppThemeContext);
