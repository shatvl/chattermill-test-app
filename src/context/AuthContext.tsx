import React, {
  FC, createContext, useContext, useState, useEffect,
} from 'react';
import jwt_decode from 'jwt-decode';
import { CHATTER_TOKEN_COOKIE, getCoockie } from '../services/cookie';
import { User } from '../models/User.model';

interface AuthContextProps {
  user: User;
  isAuthorized: () => boolean;
}

export const AuthContext = createContext<AuthContextProps>({
  user: null,
  isAuthorized: () => false,
});

export const AuthProvider: FC = ({ children }) => {
  const [user, setUser] = useState<User>();

  // decode token from params or try to retrieve from cookie
  const getUser = (token?: string): User => {
    const tok = token ?? getCoockie(CHATTER_TOKEN_COOKIE);

    if (!tok) {
      return null;
    }

    try {
      const decodedUser: { exp: number; id: string; orig_iat: number } = jwt_decode(tok);

      return decodedUser || null;
    } catch (error) {
      console.error(error);
      return null;
    }
  };

  const isAuthorized = () =>
    // check via getUser() method, because token could be expired and if users won't decoded navigate to login page
    !!getUser();
  useEffect(() => setUser(getUser()), []);

  return <AuthContext.Provider value={{ user, isAuthorized }}>{children}</AuthContext.Provider>;
};

export const useAuthContext = () => useContext(AuthContext);
