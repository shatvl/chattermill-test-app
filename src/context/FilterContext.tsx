import React, {
  FC, createContext, useContext, useState,
} from 'react';

interface FilterContextProps {
  themeId: number;
  setThemeId: (themeId: number) => void;
  clearFilters: () => void;
}

export const FilterContext = createContext<FilterContextProps>({
  themeId: null,
  setThemeId: () => {},
  clearFilters: () => {},
});

export const FilterProvider: FC = ({ children }) => {
  const [themeId, setThemeId] = useState<number>();

  const handleSelectThemeId = (themeId: number) => setThemeId(themeId);

  const handleClearFilters = () => setThemeId(null);

  return (
    <FilterContext.Provider value={{ themeId, setThemeId: handleSelectThemeId, clearFilters: handleClearFilters }}>
      {children}
    </FilterContext.Provider>
  );
};

export const useFilterContext = () => useContext(FilterContext);
