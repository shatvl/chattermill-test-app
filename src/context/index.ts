export { AppThemeContext } from './AppThemeContext';
export { usePortalAppThemeContext } from './AppThemeContext';
