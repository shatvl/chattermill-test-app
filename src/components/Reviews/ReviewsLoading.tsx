import React from 'react';
import ReviewItemLoading from './ReviewItem/ReviewItemLoading';
import { ReviewList, ReviewListItem } from './ReviewsStyled';

const ReviewsLoading = () => (
  <ReviewList>
    <ReviewListItem>
      <ReviewItemLoading />
    </ReviewListItem>
    <ReviewListItem>
      <ReviewItemLoading />
    </ReviewListItem>
    <ReviewListItem>
      <ReviewItemLoading />
    </ReviewListItem>
    <ReviewListItem>
      <ReviewItemLoading />
    </ReviewListItem>
    <ReviewListItem>
      <ReviewItemLoading />
    </ReviewListItem>
  </ReviewList>
);

export default ReviewsLoading;
