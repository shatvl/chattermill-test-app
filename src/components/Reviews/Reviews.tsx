import React, { FC, useEffect, useRef, useState } from 'react';
import { Review } from '../../models/Review.model';
import { useFeedListLazyQuery } from '../../services/review/useFeedListLazyQuery';
import InfiniteScroll from '../InfiniteScroll';
import ReviewItem from './ReviewItem';
import ThemesFilter from '../ThemesFilter';
import ReviewsLoading from './ReviewsLoading';
import { NotFoundContent, ReviewList, ReviewListItem } from './ReviewsStyled';
import { useFilterContext } from '../../context/FilterContext';
import { PrimaryButton } from '../Button';
import { ButtonSizes } from '../Button/Button';

const PAGE_SIZE = 20;

const Reviews: FC = () => {
  const [paginatedReviews, setPaginatedReviews] = useState<Review[]>([]);
  const page = useRef<number>(0);
  const [hasMore, setHasMore] = useState<boolean>(true);
  const [load, { data, loading, error }] = useFeedListLazyQuery();
  const { themeId, clearFilters } = useFilterContext();

  useEffect(() => {
    if (data && data.length) {
      setPaginatedReviews([...paginatedReviews, ...data]);
    }

    if (data && data.length % PAGE_SIZE === 0) {
      setHasMore(true);
      return;
    }

    if (data && (data.length < PAGE_SIZE || !data.length)) {
      setHasMore(false);
      return;
    }
  }, [data]);

  useEffect(() => {
    // reset old data
    setPaginatedReviews([]);
    page.current = 0;
    load({ limit: PAGE_SIZE, offset: page.current * PAGE_SIZE, themeId });
  }, [themeId]);

  const handleLoadInfinite = () => {
    page.current++;
    load({ limit: PAGE_SIZE, offset: page.current * PAGE_SIZE, themeId });
  };

  if (loading && !page.current) {
    return (
      <>
        <ThemesFilter />
        <ReviewsLoading />
      </>
    );
  }

  if (error) {
    return <div>{error}</div>;
  }

  if (!paginatedReviews.length) {
    return (
      <>
        <ThemesFilter />
        <NotFoundContent>
          No reviews found, try to{' '}
          <PrimaryButton onClick={clearFilters} size={ButtonSizes.XS}>
            clear filters
          </PrimaryButton>
        </NotFoundContent>
      </>
    );
  }

  return (
    !!paginatedReviews?.length && (
      <>
        <ThemesFilter />
        <InfiniteScroll loading={loading} hasMore={hasMore} load={handleLoadInfinite}>
          <ReviewList>
            {paginatedReviews.map(review => (
              <ReviewListItem key={review.id}>
                <ReviewItem review={review} />
              </ReviewListItem>
            ))}
          </ReviewList>
        </InfiniteScroll>
      </>
    )
  );
};

export default Reviews;
