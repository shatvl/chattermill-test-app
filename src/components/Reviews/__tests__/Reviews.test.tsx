/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

import { waitRenderer } from '../../../testing/wait';
import { AppThemeProvider } from '../../../context/AppThemeContext';
import mockFetch from '../../../testing/mockFetch';
import Reviews from '../Reviews';
import mockedResponse from '../__mocks__/response.json';

describe('renders reviews', () => {
  const mockIntersectionObserver = jest.fn();
  const mockObserve = jest.fn();
  const mockDisconnect = jest.fn();

  beforeEach(() => {
    mockIntersectionObserver.mockReturnValue({
      observe: mockObserve,
      unobserve: () => null,
      disconnect: mockDisconnect,
    });
    window.IntersectionObserver = mockIntersectionObserver;
  });

  it('reviews snapshot', async () => {
    const tree = renderer.create(
      <AppThemeProvider>
        <Reviews />
      </AppThemeProvider>,
    );

    mockFetch(200, mockedResponse);

    await waitRenderer();

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
