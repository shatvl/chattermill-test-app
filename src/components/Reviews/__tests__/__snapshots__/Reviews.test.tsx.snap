// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders reviews reviews snapshot 1`] = `
Array [
  .c3 {
  display: inline-block;
  width: initial;
  vertical-align: middle;
  background: transparent;
  border: 1px solid transparent;
  font-size: 0.875rem;
  line-height: 1rem;
  padding: 0.25rem 0.375rem;
  border-radius: 0.25rem;
  color: #ff8489;
  box-sizing: border-box;
  outline: none;
  cursor: pointer;
  -webkit-transition: 0.15s all ease-in-out;
  transition: 0.15s all ease-in-out;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
}

.c3:hover,
.c3:active {
  outline: none;
}

.c3:active {
  background-color: rgba(255,132,136,.5);
  border-color: rgba(255,132,136,.2);
}

.c3:disabled,
.c3:disabled:hover,
.c3:disabled:focus,
.c3:disabled:active {
  cursor: not-allowed;
  color: #c4c9cc;
}

.c6 .c2 {
  margin-left: -0.375rem;
}

.c0 {
  position: relative;
  max-width: 40rem;
  margin: auto;
}

.c4 {
  display: block;
  width: 100%;
  height: 3.375rem;
  padding: calc(.75rem - 0.0625rem) calc(.9rem - 0.0625rem);
  border: 0.0625rem solid #c4c9cc;
  border-radius: 0.25rem;
  outline: none;
  -webkit-transition: border-color 0.15s ease;
  transition: border-color 0.15s ease;
  word-break: break-word;
  font-size: 1rem;
  line-height: 1.25rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c5 .c2 {
  margin-left: 0.5rem;
}

<div
    className="c0"
  >
    <div
      className="c1"
    >
      <h4>
        Filter by Theme
      </h4>
      <button
        className="c2 c3"
        onClick={[Function]}
        size="xs"
      >
        Log Out
      </button>
    </div>
    <input
      className="c4"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Type theme name..."
      value=""
    />
  </div>,
  .c0 {
  overflow-y: auto;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  box-shadow: 0px 2px 8px -2px rgba(0,0,0,0.18);
  border: 1px solid e3e3e3;
  border-radius: 0.25rem;
  max-width: 40rem;
  margin: auto;
  padding: 0.5rem 1rem;
}

.c5 {
  text-align: left;
  font-size: 1rem;
  margin: 0;
  line-height: 1.125rem;
}

.c4 {
  width: 100%;
  text-align: end;
  color: #bfbfbf;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c6 .c12 {
  margin-left: -0.375rem;
}

.c7 {
  max-height: initial;
  line-height: 1.125rem;
  width: 100%;
  overflow: hidden;
}

.c10 {
  font-size: 1em;
  display: inline-block;
  height: 1.25rem;
  width: 5rem;
  max-width: 5rem;
  position: relative;
  border-radius: 0.25rem;
  background: #f2f2f2;
  overflow: hidden;
  z-index: -1;
}

.c10::before {
  content: '';
  display: block;
  position: absolute;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  -webkit-animation: imJtFj 1.2s linear infinite;
  animation: imJtFj 1.2s linear infinite;
  background-image: linear-gradient( 90deg, rgba(248,248,249,0) 0, rgba(248,248,249,1) 40%, rgba(248,248,249,1) 60%, rgba(248,248,249,0) );
}

.c11 {
  border-radius: 0.25rem;
  height: 2rem;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 0.75rem 0 0;
  padding-right: 1.5rem;
}

.c1 {
  padding: 0;
  margin: 0;
  list-style: none;
  margin-top: 1.25rem;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 0 0 1rem;
}

<div
    className="c0"
  >
    <ul
      className="c1"
    >
      <li
        className="c2"
      >
        <article
          className="c3"
        >
          <div
            className="c4"
          >
            Jul 18, 2019
          </div>
          <div
            className="c5"
          >
            <div
              className="c6"
            >
              <div
                className="c7"
              >
                It's a decent banking app. Mostly easy to use and generally reliable. Does tend to pop up more 'offers' than strictly necessary.
              </div>
            </div>
          </div>
          <div
            className="c8"
          >
            <dl
              className="c9"
            >
              <span
                className="c10 c11"
                height={2}
              />
            </dl>
            <dl
              className="c9"
            >
              <span
                className="c10 c11"
                height={2}
              />
            </dl>
            <dl
              className="c9"
            >
              <span
                className="c10 c11"
                height={2}
              />
            </dl>
          </div>
        </article>
      </li>
      <li
        className="c2"
      >
        <article
          className="c3"
        >
          <div
            className="c4"
          >
            Jul 18, 2019
          </div>
          <div
            className="c5"
          >
            <div
              className="c6"
            >
              <div
                className="c7"
              >
                statement should be available to print from app.massaging service is poor.really need a email client to talk.because it is more useful and reliable for any type of enquiry.
              </div>
            </div>
          </div>
          <div
            className="c8"
          >
            <dl
              className="c9"
            >
              <span
                className="c10 c11"
                height={2}
              />
            </dl>
            <dl
              className="c9"
            >
              <span
                className="c10 c11"
                height={2}
              />
            </dl>
          </div>
        </article>
      </li>
      <li
        className="c2"
      >
        <article
          className="c3"
        >
          <div
            className="c4"
          >
            Jul 17, 2019
          </div>
          <div
            className="c5"
          >
            <div
              className="c6"
            >
              <div
                className="c7"
              >
                It's essential for me to tracking balance every day and transfer money. I'm more and reliant Internet banking it's instantaneous trusted most of all very excellent banking site.
              </div>
            </div>
          </div>
          <div
            className="c8"
          >
            <dl
              className="c9"
            >
              <span
                className="c10 c11"
                height={2}
              />
            </dl>
            <dl
              className="c9"
            >
              <span
                className="c10 c11"
                height={2}
              />
            </dl>
            <dl
              className="c9"
            >
              <span
                className="c10 c11"
                height={2}
              />
            </dl>
            <dl
              className="c9"
            >
              <span
                className="c10 c11"
                height={2}
              />
            </dl>
          </div>
        </article>
      </li>
      <li
        className="c2"
      >
        <article
          className="c3"
        >
          <div
            className="c4"
          >
            Jul 17, 2019
          </div>
          <div
            className="c5"
          >
            <div
              className="c6"
            >
              <div
                className="c7"
              >
                Very Reliable
              </div>
            </div>
          </div>
          <div
            className="c8"
          >
            <dl
              className="c9"
            >
              <span
                className="c10 c11"
                height={2}
              />
            </dl>
          </div>
        </article>
      </li>
    </ul>
    <div
      className=""
    />
  </div>,
]
`;
